keys.go

```diff
@@ -0,0 +1,174 @@
+	HasTLSFingerprint         bool // The response includes a TLS fingerprint.
+	ValidSHA256TLSFingerprint bool // Every TLS fingerprint includes a SHA-256 hash.
+	MatchingTLSFingerprint    bool // The TLS fingerprint for the connection matches one of the listed fingerprints.
+}
+
+// CheckKeys checks the keys returned from a server to make sure they are valid.
+// If the checks pass the also returns a map of key_id to Ed25519 public key and a list of SHA256 TLS fingerprints.
+func CheckKeys(serverName string, timeNowMs uint64, keys ServerKeys, connState *tls.ConnectionState) (
+	checks KeyChecks, ed25519Keys map[string]Base64String, sha256Fingerprints []Base64String,
+) {
+	checks.MatchingServerName = serverName == keys.ServerName
+
+	ed25519Keys = checkVerifyKeys(keys, &checks)
+	sha256Fingerprints = checkTLSFingerprints(keys, &checks)
+
+	checks.FutureValidUntilTS = timeNowMs < keys.ValidUntilTS
+
+	if connState == nil {
+		// No TLS connection state so skip the fingerprint checks.
+		checks.MatchingTLSFingerprint = true
+	} else {
+		// Check the peer certificates.
+		checks.MatchingTLSFingerprint = checkFingerprint(connState, sha256Fingerprints)
+	}
+
+	checks.AllChecksOK = (checks.MatchingServerName &&
+		checks.FutureValidUntilTS &&
+		checks.HasEd25519VerifyKey &&
+		checks.ValidEd25519VerifyKey &&
+		checks.MatchingEd25519Signature &&
+		checks.HasTLSFingerprint &&
+		checks.ValidSHA256TLSFingerprint &&
+		checks.MatchingTLSFingerprint)
+
+	if !checks.AllChecksOK {
+		sha256Fingerprints = nil
+		ed25519Keys = nil
+	}
+	return
+}
+
+func checkFingerprint(connState *tls.ConnectionState, sha256Fingerprints []Base64String) bool {
+	if len(connState.PeerCertificates) == 0 {
+		return false
+	}
+	cert := connState.PeerCertificates[0]
+	digest := sha256.Sum256(cert.Raw)
+	for _, fingerprint := range sha256Fingerprints {
+		if bytes.Compare(digest[:], fingerprint) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func checkVerifyKeys(keys ServerKeys, checks *KeyChecks) map[string]Base64String {
+	checks.ValidEd25519VerifyKey = true
+	checks.MatchingEd25519Signature = true
+	verifyKeys := map[string]Base64String{}
+	for keyID, keyData := range keys.VerifyKeys {
+		algorithm := strings.SplitN(keyID, ":", 2)[0]
+		publicKey := keyData.Key
+		if algorithm == "ed25519" {
+			if len(publicKey) != 32 {
+				checks.ValidEd25519VerifyKey = false
+				continue
+			}
+			if err := VerifyJSON(keys.ServerName, keyID, []byte(publicKey), keys.Raw); err != nil {
+				checks.MatchingEd25519Signature = false
+				continue
+			}
+			checks.HasEd25519VerifyKey = true
+			verifyKeys[keyID] = publicKey
+		}
+	}
+	return verifyKeys
+}
+
+func checkTLSFingerprints(keys ServerKeys, checks *KeyChecks) []Base64String {
+	var fingerprints []Base64String
+	checks.ValidSHA256TLSFingerprint = true
+	for _, fingerprint := range keys.TLSFingerprints {
+		if len(fingerprint.SHA256) != sha256.Size {
+			checks.ValidSHA256TLSFingerprint = false
```

Comment:
Seems like a bit of a weak check. Isn't there something better we can do aside from a basic length check?